### ISSUE TITLE
Enforce the upwards data flow interface

### ISSUE THREAD

**Issue by kaste** [Polymer/polymer#3553]
Allow sending arbitrary name/value pairs upwards via `this.fire('name-changed', value)` and let them participate in the data flow.

Two use-cases for this: 
1. Generally any data/event producer could just forward or pass new data (from a stream) upwards without storing it locally. Consider `<iron-ajax>`, instead of `this.response = ...` we would just `notifyUpwards('response', ...)` and forget about the response immediately.
2. The producing element just produces arbitrary name/value pairs that cannot be hardcoded or foreseen at all. The user uses this element like every other element. Consider a typical route-matching element, based on user input (the route pattern) varying symbols (in the example below userId or postId) will pop up. So the element gets a _polymorphic_ shape.

```
    <x-ample 
        pattern="/user/:userId" 
        url="[[location]]" 
        userId="{{userId}}"
    ></x-ample>
    <x-ample 
        pattern="/blog/:postId" 
        url="[[location]]" 
        postId="{{postId}}"
    ></x-ample>
    <dom-module id="x-ample">
        <script>
            Polymer({
                is: 'x-ample',
                properties: {
                    pattern: String,
                    url: String,
                },
                observers: ['match(pattern, url)'],
                match: function(pattern, url) {
                    /*  m a g i c */
                    for (var name in matches) {
                        this.notifyUp(name, matches[name]);
                    }
                }
            });
        </script>
    <dom-module>
```

**Comment by TimvdLippe** [comment]
In your example, where is `matches` defined?

**Comment by kaste** [comment]
In `/*  m a g i c */`.

**Comment by kaste** [comment]
This is essentially blocked by a suspicious comment by @sorvell https://github.com/Polymer/polymer/blob/edb59eb6283074ce54f374b30b884c9f9fcf9f5f/src/lib/bind/accessors.html#L222

>  // TODO(sorvell): even though we have a `value` argument, we *must*
>  // lookup the current value of the property. Multiple listeners and
>  // queued events during configuration can theoretically lead to
>  // divergence of the passed value from the current value, but we
>  // really need to track down a specific case where this happens.

The problem that lead to this comment was reported by me Oct 2015 (!). 

Note that we don't have a test that covers the above comment. What I usually would do is, remove that comment, fix the code, and wait for real, actionable bug reports. Life goes on. (Since this is related to config-phase the new fixes would go into configure.html not into accessors.html.)

That said I was able to produce a test case by thought experiments that leads to said divergence of the passed value. But I had to write a nonsensical tree of custom elements. The final outcome/value that survived the side-effects run was unexpected, and flaky b/c it relied completely on the kind and thus order of effects. So even though I saw a divergence between the passed and the current value, the divergence is not the problem here, the problem is what is the _correct_ value the program should use, the user intends to survive. My perspective here is that it seems undecidable. (I wrote this first pencil-paper, and promptly expected the passed value would survive.)
